extensions/amp-sticky-ad/0.1/amp-sticky-ad.js

```diff
@@ -37,6 +38,72 @@ class AmpStickyAd extends AMP.BaseElement {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    const children = this.getRealChildren();
+    user.assert((children.length == 1 && children[0].tagName == 'AMP-AD'),
+        'amp-sticky-ad must have a single amp-ad child');
+    /** @const @private {!Element} */
+    this.ad_ = children[0];
+
+    /** @private @const {!Viewport} */
+    this.viewport_ = this.getViewport();
+
+    /** @const @private {!Vsync} */
+    this.vsync_ = this.getVsync();
+
+    /** @const @private {!Unlisten}
+     * On viewport scroll, check requirements for amp-stick-ad to display.
+     */
+    this.onScrollListener_ =
+        this.viewport_.onScroll(() => this.displayAfterScroll());
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
+    if (!this.isExperimentOn_) {
+      dev.warn(TAG, `TAG ${TAG} disabled`);
+      return Promise.resolve();
+    }
+    return Promise.resolve();
+  }
+
+  /** @private
+   * The listener function that listen on onScroll event and
+   * show sticky ad when user scroll at least one viewport and
+   * there is at least one more viewport available.
+   */
+  displayAfterScroll() {
+    this.scrollTop_ = this.viewport_.getScrollTop();
+    this.scrollHeight_ = this.viewport_.getScrollHeight();
+    this.viewportHeight_ = this.viewport_.getSize().height;
+    // TODO(zhouyx): When calculate 'has scrolled through at least 1 viewport'
+    // do we want to count height from top or from initial position?
+    // Will assume user start from `this.scrollTop_ = 0` here. Need
+    // to figure out later.
+
+    // Check user has scrolled at least one viewport from init position.
+    if (this.viewportHeight_ < this.scrollTop_) {
+      const remainHeight = this.scrollHeight_
+          - this.scrollTop_ - this.viewportHeight_;
+      if (remainHeight < this.viewportHeight_) {
+        // TODO(zhouyx): Figure if early unlisten is needed earlier
+        // if scrollHeight is less than 2*viewportHeight.
+        this.onScrollListener_();
```

Comment:
Let's put the unlisten operation into it's own method. Besides the two places here we also need to call it on element's removal from DOM. This callback arrives through `custom-element.js ElementProto.detachedCallback`. Unfortunately we currently do not expose this callback to `BaseElement` so you'll have to propagate it through.